src/runtime/micro/micro_device_api.cc

```diff
@@ -0,0 +1,173 @@
+ */
+class MicroDeviceAPI final : public DeviceAPI {
+ public:
+  /*! \brief constructor */
+  MicroDeviceAPI() { }
+
+  void SetDevice(TVMContext ctx) final {}
+
+  void GetAttr(TVMContext ctx, DeviceAttrKind kind, TVMRetValue* rv) final {
+    if (kind == kExist) {
+      *rv = 1;
+    }
+  }
+
+  void* AllocDataSpace(TVMContext ctx,
+                       size_t nbytes,
+                       size_t alignment,
+                       TVMType type_hint) final {
+    std::shared_ptr<MicroSession> session = MicroSession::Global();
+    void* data = session->AllocateInSection(SectionKind::kHeap, nbytes).cast_to<void*>();
+    CHECK(data != nullptr) << "unable to allocate " << nbytes << " bytes on device heap";
+    MicroDevSpace* dev_space = new MicroDevSpace();
+    dev_space->data = data;
+    dev_space->session = session;
+    return static_cast<void*>(dev_space);
+  }
+
+  void FreeDataSpace(TVMContext ctx, void* ptr) final {
+    MicroDevSpace* dev_space = static_cast<MicroDevSpace*>(ptr);
+    dev_space->session->FreeInSection(
+      SectionKind::kHeap, DevBaseOffset(reinterpret_cast<std::uintptr_t>(dev_space->data)));
+    delete dev_space;
+  }
+
+  void CopyDataFromTo(const void* from,
+                      size_t from_offset,
+                      void* to,
+                      size_t to_offset,
+                      size_t size,
+                      TVMContext ctx_from,
+                      TVMContext ctx_to,
+                      TVMType type_hint,
+                      TVMStreamHandle stream) final {
+    std::tuple<int, int> type_from_to(ctx_from.device_type, ctx_to.device_type);
+    if (type_from_to == std::make_tuple(kDLMicroDev, kDLMicroDev)) {
+      // Copying from the device to the device.
+
+      MicroDevSpace* from_space = static_cast<MicroDevSpace*>(const_cast<void*>(from));
+      MicroDevSpace* to_space = static_cast<MicroDevSpace*>(const_cast<void*>(to));
+      CHECK(from_space->session == to_space->session)
+          << "attempt to copy data between different micro sessions (" << from_space->session
+          << " != " << to_space->session << ")";
+      CHECK(ctx_from.device_id == ctx_to.device_id)
+        << "can only copy between the same micro device";
+      std::shared_ptr<MicroSession> session = from_space->session;
+      const std::shared_ptr<LowLevelDevice>& lld = session->low_level_device();
+
+      DevBaseOffset from_dev_offset = GetDevLoc(from_space, from_offset);
+      DevBaseOffset to_dev_offset = GetDevLoc(to_space, to_offset);
+
+      std::vector<uint8_t> buffer(size);
+      lld->Read(from_dev_offset, static_cast<void*>(buffer.data()), size);
+      lld->Write(to_dev_offset, static_cast<void*>(buffer.data()), size);
+    } else if (type_from_to == std::make_tuple(kDLMicroDev, kDLCPU)) {
+      // Reading from the device.
+
+      MicroDevSpace* from_space = static_cast<MicroDevSpace*>(const_cast<void*>(from));
+      std::shared_ptr<MicroSession> session = from_space->session;
```

Comment:
use reference instead of deep copy, this will increase the reference counter of the session, can inline into the next line